File: src/main/java/pw/switchcraft/plethora/gameplay/modules/glasses/canvas/CanvasClient.java
package pw.switchcraft.plethora.gameplay.modules.glasses.canvas;

import it.unimi.dsi.fastutil.ints.*;
import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.minecraft.client.util.math.MatrixStack;
import pw.switchcraft.plethora.Plethora;
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.BaseObject;
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.ObjectGroup;

import static pw.switchcraft.plethora.gameplay.modules.glasses.canvas.CanvasHandler.ID_2D;
import static pw.switchcraft.plethora.gameplay.modules.glasses.canvas.CanvasHandler.ID_3D;

public class CanvasClient {
    public final int id;

    private final Int2ObjectMap<BaseObject> objects = new Int2ObjectOpenHashMap<>();
    private final Int2ObjectMap<IntSortedSet> childrenOf = new Int2ObjectOpenHashMap<>();

    public CanvasClient(int id) {
        this.id = id;
        childrenOf.put(ID_2D, new IntAVLTreeSet());
        childrenOf.put(ID_3D, new IntAVLTreeSet());
    }

    public void updateObject(BaseObject object) {
        IntSet parent = childrenOf.get(object.parent());
        if (parent == null) {
            Plethora.LOG.error("Trying to add " + object.id() + " to group " + object.parent() + " (" + object + ")");
            return; // Should never happen but...
        }

        if (objects.put(object.id(), object) == null) {
            // If this is a new instance then set up the children
            parent.add(object.id());
            if (object instanceof ObjectGroup) childrenOf.put(object.id(), new IntAVLTreeSet());
        }
    }

    public void remove(int id) {
        BaseObject object = objects.remove(id);
        childrenOf.remove(id); // We handle the removing of children in the canvas version

        if (object != null) {
            // Remove from the parent set if needed.
            IntSet parent = childrenOf.get(object.parent());
            if (parent != null) parent.remove(id);
        }
    }

    public BaseObject getObject(int id) {
        return objects.get(id);
    }

    public IntSet getChildren(int id) {
        return childrenOf.get(id);
    }

    @Environment(EnvType.CLIENT)
    public void drawChildren(IntIterator children, MatrixStack matrices) {
        while (children.hasNext()) {
            int id = children.nextInt();
            BaseObject object = getObject(id);
            if (object != null) object.draw(this, matrices);
        }
    }
}

File: src/main/java/io/sc3/plethora/gameplay/modules/glasses/canvas/CanvasClient.kt
package io.sc3.plethora.gameplay.modules.glasses.canvas

import it.unimi.dsi.fastutil.ints.*
import it.unimi.dsi.fastutil.ints.IntIterator
import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.minecraft.client.render.VertexConsumerProvider
import net.minecraft.client.util.math.MatrixStack
import io.sc3.plethora.Plethora
import io.sc3.plethora.gameplay.modules.glasses.objects.BaseObject
import io.sc3.plethora.gameplay.modules.glasses.objects.ObjectGroup

class CanvasClient(val id: Int) {
  private val objects: Int2ObjectMap<BaseObject?> = Int2ObjectOpenHashMap()
  private val childrenOf: Int2ObjectMap<IntSortedSet> = Int2ObjectOpenHashMap()

  init {
    childrenOf.put(CanvasHandler.ID_2D, IntAVLTreeSet())
    childrenOf.put(CanvasHandler.ID_3D, IntAVLTreeSet())
  }

  fun updateObject(obj: BaseObject) {
    val parent: IntSet? = childrenOf[obj.parent]
    if (parent == null) {
      Plethora.log.error("Trying to add ${obj.id} to group ${obj.parent} ($obj)")
      return  // Should never happen but...
    }

    if (objects.put(obj.id, obj) == null) {
      // If this is a new instance then set up the children
      parent.add(obj.id)
      if (obj is ObjectGroup) childrenOf.put(obj.id, IntAVLTreeSet())
    }
  }

  fun remove(id: Int) {
    val obj = objects.remove(id)
    childrenOf.remove(id) // We handle the removing of children in the canvas version

    if (obj != null) {
      // Remove from the parent set if needed.
      val parent: IntSet? = childrenOf[obj.parent]
      parent?.remove(id)
    }
  }

  fun getObject(id: Int) = objects[id]

  fun getChildren(id: Int): IntSet? {
    val children = childrenOf[id]
    return if (children.isEmpty()) null else children
  }

  @Environment(EnvType.CLIENT)
  fun drawChildren(children: IntIterator, matrices: MatrixStack?, consumers: VertexConsumerProvider?) {
    while (children.hasNext()) {
      val id = children.nextInt()
      val obj = getObject(id)
      obj?.draw(this, matrices!!, consumers)
    }
  }
}
